### ISSUE TITLE
using the json3/srv3 format and use the gaps and confidence ratings in the prediction

### ISSUE THREAD

**Issue by nikitalita** [shashank2123/Punctuation-Restoration-For-Youtube-Transcript#1]
Hey, I was interested in doing something exactly like this!

I was wondering if you would be interested in adding the gap duration between words in the transcript to further determine whether or not to add a sentence or utterance stop

Youtube provides this information: In addition to the standard VTT subtitles, youtube also offers a json3 formatted file that contains the exact time the word is uttered and its confidence score (from 0-255, it seems). You can see the json3 format for yourself if you go to a youtube video with an autogenerated transcript, open developer tools, click the CC button, and open up the timedtext url in the developer tools network tab. (youtube-transcript-api also has an issue talking about adding the ability to download these here: https://github.com/jdepoix/youtube-transcript-api/issues/106)

[Example json3 file](https://pastebin.com/46kai2p9)

As for using this info (outside of a pytorch model), what you could possibly do is, after running it through pytorch, decide whether or not to turn a comma into a period, or, if there's a large gap between words that wasn't previously predicted to be a sentence end or sentence particle, add an utterence stop ('--')

Additionally, you could use the prediction score to determine if there are two people talking over each other in the transcript. I usually notice that the confidence score takes a dive in the middle of the sequence where people are talking over each other; you could potentially add utterence stops between these sentences (or flag them for human evaluation).

**Comment by nikitalita** [comment]
I also note that https://github.com/xashru/punctuation-restoration tends to be a bit more accurate in determining when to add periods instead of commas, so you may want to take a look at that.

**Comment by shashank2123** [comment]
> I also note that https://github.com/xashru/punctuation-restoration tends to be a bit more accurate in determining when to add periods instead of commas, so you may want to take a look at that.

Yeah, I know  I saw  https://github.com/xashru/punctuation-restoration there they used Bert model like how we predict Named Entity Recognitions task that was great idea. In my work there is a issue of missing word in validation dataset when I am calculating accuracy or other metrics  I comparing actual sentence with missing predictions so that might effect the result. Another thing is I ran model just for 3 epochs So I will try to improve the performance.

**Comment by shashank2123** [comment]
> Hey, I was interested in doing something exactly like this!
> 
> I was wondering if you would be interested in adding the gap duration between words in the transcript to further determine whether or not to add a sentence or utterance stop
> 
> Youtube provides this information: In addition to the standard VTT subtitles, youtube also offers a json3 formatted file that contains the exact time the word is uttered and its confidence score (from 0-255, it seems). You can see the json3 format for yourself if you go to a youtube video with an autogenerated transcript, open developer tools, click the CC button, and open up the timedtext url in the developer tools network tab. (youtube-transcript-api also has an issue talking about adding the ability to download these here: [jdepoix/youtube-transcript-api#106](https://github.com/jdepoix/youtube-transcript-api/issues/106))
> 
> [Example json3 file](https://pastebin.com/46kai2p9)
> 
> As for using this info (outside of a pytorch model), what you could possibly do is, after running it through pytorch, decide whether or not to turn a comma into a period, or, if there's a large gap between words that wasn't previously predicted to be a sentence end or sentence particle, add an utterence stop ('--')
> 
> Additionally, you could use the prediction score to determine if there are two people talking over each other in the transcript. I usually notice that the confidence score takes a dive in the middle of the sequence where people are talking over each other; you could potentially add utterence stops between these sentences (or flag them for human evaluation).

Hey , sorry I didn't get everything but I have some doubts when I looked into json3 formatted offered by youtube I didn't get what is the meaning some key terms. could you please explain me what that means?
You told about adding the gap duration between words and further determine whether or not to add a sentence or utterance stop could please tell with some examples? [Sorrry for bothering you]

**Comment by nikitalita** [comment]
Sure. Consider the following sequence:

```json
    {
    "tStartMs": 80,
    "dDurationMs": 3680,
    "wWinId": 1,
    "segs": [ {
      "utf8": "since",
      "acAsrConf": 255
    }, 
    {
      "utf8": " you",
      "tOffsetMs": 319,
      "acAsrConf": 255
    }, {
      "utf8": " guys",
      "tOffsetMs": 400,
      "acAsrConf": 255
    }, {
      "utf8": " love",
      "tOffsetMs": 719,
      "acAsrConf": 255
    }, {
      "utf8": " my",
      "tOffsetMs": 960,
      "acAsrConf": 255
    }, {
      "utf8": " thrifting",
      "tOffsetMs": 1119,
      "acAsrConf": 255
    }, {
      "utf8": " videos",
      "tOffsetMs": 1519,
      "acAsrConf": 255
    } ]
  }, {
    "tStartMs": 2070,
    "dDurationMs": 1690,
    "wWinId": 1,
    "aAppend": 1,
    "segs": [ {
      "utf8": "\n"
    } ]
  }, {
    "tStartMs": 2080,
    "dDurationMs": 3440,
    "wWinId": 1,
    "segs": [ {
      "utf8": "so",
      "acAsrConf": 255
    }, {
      "utf8": " much",
      "tOffsetMs": 240,
      "acAsrConf": 255
    }, {
      "utf8": " in",
      "tOffsetMs": 720,
      "acAsrConf": 255
    }, {
      "utf8": " today's",
      "tOffsetMs": 799,
      "acAsrConf": 255
    }, {
      "utf8": " video",
      "tOffsetMs": 1119,
      "acAsrConf": 255
    } ]
```

The json3 transcript is broken up into a sequence of segments (`segs`), each sequence containing the start time of the sequence (`tStartMs`) in milliseconds since the beginning of the video, and the duration in milliseconds that the sequence is going to be on the screen for (`dDurationMs`) (ignore wWinId for now).
```json
    {
    "tStartMs": 80,
    "dDurationMs": 3680,
    "wWinId": 1,
    "segs": [
```

For each word segment in a segment sequence, it contains the offset (`tOffsetMs` ) relative to `tStartMs`; so, for this example, since "since" has no offset, it would display at 80ms, "you" would display at 399ms, etc.
Additonally, each word segment has a confidence rating (`acAsrConf`) which is an indicator of how confident Youtube's speech-to-text engine was that the word they interpreted is accurate.
```json
    "segs": [ {
      "utf8": "since",
      "acAsrConf": 255
    }, {
      "utf8": " you",
      "tOffsetMs": 319,
      "acAsrConf": 255
    }, {
      "utf8": " guys",
      "tOffsetMs": 400,
      "acAsrConf": 255
    }, {
      "utf8": " love",
      "tOffsetMs": 719,
      "acAsrConf": 255
    }, {
      "utf8": " my",
      "tOffsetMs": 960,
      "acAsrConf": 255
    }, {
      "utf8": " thrifting",
      "tOffsetMs": 1119,
      "acAsrConf": 255
    }, {
      "utf8": " videos",
      "tOffsetMs": 1519,
      "acAsrConf": 255
    } ]
```

as for the gaps, consider the next segment sequence:

```json
  }, {
    "tStartMs": 2080,
    "dDurationMs": 3440,
    "wWinId": 1,
    "segs": [ {
      "utf8": "so",
      "acAsrConf": 255
    }, {
      "utf8": " much",
      "tOffsetMs": 240,
      "acAsrConf": 255
    }, {
      "utf8": " in",
      "tOffsetMs": 720,
      "acAsrConf": 255
    }, {
      "utf8": " today's",
      "tOffsetMs": 799,
      "acAsrConf": 255
    }, {
      "utf8": " video",
      "tOffsetMs": 1119,
      "acAsrConf": 255
    } ]
```

If you look between "much" and "in, there's a large enough gap between these words (480ms, almost half a second) to determine that there should be a comma.

```json
    "segs": [ {
      "utf8": "for",
      "acAsrConf": 255
    }, {
      "utf8": " 2021",
      "tOffsetMs": 400,
      "acAsrConf": 255
    }, {
      "utf8": " trends",
      "tOffsetMs": 1120,
      "acAsrConf": 255
    }, {
      "utf8": " i'm",
      "tOffsetMs": 1760,
      "acAsrConf": 255
    }, {
      "utf8": " going",
      "tOffsetMs": 1919,
      "acAsrConf": 255
    }, {
      "utf8": " to",
      "tOffsetMs": 1999,
      "acAsrConf": 255
    }, {
```

In this other sequence, there's a large gap between "trends" and "i'm", long enough to indicate that there's a period between these words. 
It also LOOKS LIKE there's a large gap between "2021" and "trends"; However, since "twenty twenty one" takes a lot longer to say than "trends", and "tOffsetMs" is the START time when the word is said rather than the end time, this is not actually an indicator that there should be a period here. So, word length and syllables need to be taken into account here.

As for "utterance stop", I mean that where there's an indication that the speaker stopped talking in the middle of a sentence before resuming; basically, when the model predicts that there was a whole sentence among a series of word segments, but there's a large gap between two of the words within it.

**Comment by shashank2123** [comment]
Thank you so much for detailed explanation I really appreciate it.  Now I saw there is  "utterance stop", Before I didn't thought to consider it [ to be frank I didn't know what its real meaning]. I will explore dataset more and consider all punctuation that are important.
As you know I am using TED-Talk dataset so I think word gaps are consider and punctuated if not I will check what they did for word gap .  
If your saying like after prediction we need to consider words gap make change as you told earlier. Ok it seem interest I will see what I can do.
Thank you so much for mentioning I never thought in this point of view.
